File: src/lib/quadtree.ts
/**
 * quadtree-js
 * @version 1.2.4
 * @license MIT
 * @author Timo Hausmann
 *
 * TypeScript conversion for classic-wgl
 */

/**
 * The Quadtree uses rectangle objects for all areas ("Rect").
 * All rectangles require the properties x, y, width, height
 */
export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

/**
 * Quadtree class for spatial partitioning
 */
export class Quadtree<T extends Rect = Rect> {
  maxObjects: number;
  maxLevels: number;
  level: number;
  bounds: Rect;
  objects: T[];
  nodes: Quadtree<T>[];

  /**
   * @param bounds - bounds of the node ({ x, y, width, height })
   * @param maxObjects - max objects a node can hold before splitting (default: 10)
   * @param maxLevels - total max levels inside root Quadtree (default: 4)
   * @param level - depth level, required for subnodes (default: 0)
   */
  constructor(
    bounds: Rect,
    maxObjects: number = 10,
    maxLevels: number = 4,
    level: number = 0
  ) {
    this.maxObjects = maxObjects;
    this.maxLevels = maxLevels;
    this.level = level;
    this.bounds = bounds;
    this.objects = [];
    this.nodes = [];
  }

  /**
   * Split the node into 4 subnodes
   */
  split(): void {
    const nextLevel = this.level + 1;
    const subWidth = this.bounds.width / 2;
    const subHeight = this.bounds.height / 2;
    const x = this.bounds.x;
    const y = this.bounds.y;

    // top right node
    this.nodes[0] = new Quadtree<T>(
      {
        x: x + subWidth,
        y: y,
        width: subWidth,
        height: subHeight,
      },
      this.maxObjects,
      this.maxLevels,
      nextLevel
    );

    // top left node
    this.nodes[1] = new Quadtree<T>(
      {
        x: x,
        y: y,
        width: subWidth,
        height: subHeight,
      },
      this.maxObjects,
      this.maxLevels,
      nextLevel
    );

    // bottom left node
    this.nodes[2] = new Quadtree<T>(
      {
        x: x,
        y: y + subHeight,
        width: subWidth,
        height: subHeight,
      },
      this.maxObjects,
      this.maxLevels,
      nextLevel
    );

    // bottom right node
    this.nodes[3] = new Quadtree<T>(
      {
        x: x + subWidth,
        y: y + subHeight,
        width: subWidth,
        height: subHeight,
      },
      this.maxObjects,
      this.maxLevels,
      nextLevel
    );
  }

  /**
   * Determine which node the object belongs to
   * @param pRect - bounds of the area to be checked
   * @returns an array of indexes of the intersecting subnodes (0-3 = top-right, top-left, bottom-left, bottom-right)
   */
  getIndex(pRect: Rect): number[] {
    const indexes: number[] = [];
    const verticalMidpoint = this.bounds.x + this.bounds.width / 2;
    const horizontalMidpoint = this.bounds.y + this.bounds.height / 2;

    const startIsNorth = pRect.y < horizontalMidpoint;
    const startIsWest = pRect.x < verticalMidpoint;
    const endIsEast = pRect.x + pRect.width > verticalMidpoint;
    const endIsSouth = pRect.y + pRect.height > horizontalMidpoint;

    // top-right quad
    if (startIsNorth && endIsEast) {
      indexes.push(0);
    }

    // top-left quad
    if (startIsWest && startIsNorth) {
      indexes.push(1);
    }

    // bottom-left quad
    if (startIsWest && endIsSouth) {
      indexes.push(2);
    }

    // bottom-right quad
    if (endIsEast && endIsSouth) {
      indexes.push(3);
    }

    return indexes;
  }

  /**
   * Insert the object into the node. If the node exceeds the capacity,
   * it will split and add all objects to their corresponding subnodes.
   * @param pRect - bounds of the object to be added
   */
  insert(pRect: T): void {
    let indexes: number[];

    // if we have subnodes, call insert on matching subnodes
    if (this.nodes.length) {
      indexes = this.getIndex(pRect);

      for (let i = 0; i < indexes.length; i++) {
        this.nodes[indexes[i]].insert(pRect);
      }
      return;
    }

    // otherwise, store object here
    this.objects.push(pRect);

    // max_objects reached
    if (this.objects.length > this.maxObjects && this.level < this.maxLevels) {
      // split if we don't already have subnodes
      if (!this.nodes.length) {
        this.split();
      }

      // add all objects to their corresponding subnode
      for (let i = 0; i < this.objects.length; i++) {
        indexes = this.getIndex(this.objects[i]);
        for (let k = 0; k < indexes.length; k++) {
          this.nodes[indexes[k]].insert(this.objects[i]);
        }
      }

      // clean up this node
      this.objects = [];
    }
  }

  /**
   * Return all objects that could collide with the given object
   * @param pRect - bounds of the object to be checked
   * @returns array with all detected objects
   */
  retrieve(pRect: Rect): T[] {
    const indexes = this.getIndex(pRect);
    let returnObjects = this.objects.slice();

    // if we have subnodes, retrieve their objects
    if (this.nodes.length) {
      for (let i = 0; i < indexes.length; i++) {
        returnObjects = returnObjects.concat(this.nodes[indexes[i]].retrieve(pRect));
      }
    }

    // remove duplicates
    returnObjects = returnObjects.filter((item, index) => {
      return returnObjects.indexOf(item) >= index;
    });

    return returnObjects;
  }

  /**
   * Clear the quadtree
   */
  clear(): void {
    this.objects = [];

    for (let i = 0; i < this.nodes.length; i++) {
      if (this.nodes.length) {
        this.nodes[i].clear();
      }
    }

    this.nodes = [];
  }
}

// Export for window global (backwards compatibility)
if (typeof window !== 'undefined') {
  (window as unknown as { Quadtree: typeof Quadtree }).Quadtree = Quadtree;
}

File: crates/classic-core/src/quadtree.rs
//! Generic spatial quadtree.
//!
//! Port of `src/lib/quadtree.ts`.

use crate::types::Rect;

/// A Quadtree node for spatial partitioning of objects that can provide a `Rect` bounding box.
pub struct Quadtree<T: RectBounds> {
    max_objects: usize,
    max_levels: usize,
    level: usize,
    bounds: Rect,
    pub objects: Vec<T>,
    pub nodes: Vec<Quadtree<T>>,
}

/// Trait for objects that can be stored in the quadtree.
pub trait RectBounds {
    fn rect(&self) -> Rect;
}

impl RectBounds for Rect {
    fn rect(&self) -> Rect {
        *self
    }
}

impl<T: RectBounds> Quadtree<T> {
    /// Create a new quadtree node.
    ///
    /// * `bounds` — the spatial bounds of this node
    /// * `max_objects` — max objects before splitting (default: 10)
    /// * `max_levels` — max depth (default: 4)
    pub fn new(bounds: Rect, max_objects: usize, max_levels: usize) -> Self {
        Self { max_objects, max_levels, level: 0, bounds, objects: Vec::new(), nodes: Vec::new() }
    }

    fn new_child(&self, bounds: Rect, level: usize) -> Self {
        Self {
            max_objects: self.max_objects,
            max_levels: self.max_levels,
            level,
            bounds,
            objects: Vec::new(),
            nodes: Vec::new(),
        }
    }

    /// Split this node into 4 sub-nodes.
    fn split(&mut self) {
        let next_level = self.level + 1;
        let sub_w = self.bounds.width / 2.0;
        let sub_h = self.bounds.height / 2.0;
        let x = self.bounds.x;
        let y = self.bounds.y;

        // 0: top-right
        self.nodes.push(self.new_child(Rect::new(x + sub_w, y, sub_w, sub_h), next_level));
        // 1: top-left
        self.nodes.push(self.new_child(Rect::new(x, y, sub_w, sub_h), next_level));
        // 2: bottom-left
        self.nodes.push(self.new_child(Rect::new(x, y + sub_h, sub_w, sub_h), next_level));
        // 3: bottom-right
        self.nodes.push(self.new_child(Rect::new(x + sub_w, y + sub_h, sub_w, sub_h), next_level));
    }

    /// Determine which sub-node indices a rectangle intersects.
    pub fn get_index(&self, r: &Rect) -> Vec<usize> {
        let mut idx = Vec::new();
        let v_mid = self.bounds.x + self.bounds.width / 2.0;
        let h_mid = self.bounds.y + self.bounds.height / 2.0;

        let start_north = r.y < h_mid;
        let start_west = r.x < v_mid;
        let end_east = r.x + r.width > v_mid;
        let end_south = r.y + r.height > h_mid;

        // top-right (0)
        if start_north && end_east {
            idx.push(0);
        }
        // top-left (1)
        if start_west && start_north {
            idx.push(1);
        }
        // bottom-left (2)
        if start_west && end_south {
            idx.push(2);
        }
        // bottom-right (3)
        if end_east && end_south {
            idx.push(3);
        }

        idx
    }

    /// Insert an object into the quadtree.
    pub fn insert(&mut self, obj: T) {
        // If we have subnodes, delegate to the one(s) that fully contain the object.
        if !self.nodes.is_empty() {
            let idxs = self.get_index(&obj.rect());
            if idxs.len() == 1 {
                self.nodes[idxs[0]].insert(obj);
            } else {
                // Straddles multiple subnodes — keep in this level.
                self.objects.push(obj);
            }
            return;
        }

        self.objects.push(obj);

        if self.objects.len() > self.max_objects && self.level < self.max_levels {
            if self.nodes.is_empty() {
                self.split();
            }
            let objs = std::mem::take(&mut self.objects);
            for obj in objs {
                let idxs = self.get_index(&obj.rect());
                if idxs.len() == 1 {
                    self.nodes[idxs[0]].insert(obj);
                } else {
                    // Straddler — stays at this level.
                    self.objects.push(obj);
                }
            }
        }
    }

    /// Retrieve all objects that could intersect with the given rectangle.
    pub fn retrieve(&self, r: &Rect) -> Vec<&T> {
        let idxs = self.get_index(r);
        let mut result: Vec<&T> = self.objects.iter().collect();

        if !self.nodes.is_empty() {
            for &i in &idxs {
                result.extend(self.nodes[i].retrieve(r));
            }
        }

        // No dedup needed — each object lives in exactly one node.
        result
    }

    /// Clear all objects and sub-nodes.
    pub fn clear(&mut self) {
        self.objects.clear();
        for node in &mut self.nodes {
            node.clear();
        }
        self.nodes.clear();
    }
}
